libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
    });
 }
 
+namespace {
+void RoundToNearestClipSample(const WaveTrack& track, double& t)
+{
+   const auto clip = track.GetClipAtTime(t);
+   if (!clip)
+      return;
+   t = clip->SamplesToTime(clip->TimeToSamples(t - clip->GetPlayStartTime())) +
+       clip->GetPlayStartTime();
+}
+}
+
+std::pair<size_t, size_t> WaveTrack::GetFloatsCenteredAroundTime(
+   double t, size_t iChannel, float* buffer, size_t numSideSamples,
+   bool mayThrow) const
+{
+   const auto numSamplesReadLeft = GetFloatsFromTime(
+      t, iChannel, buffer, numSideSamples, mayThrow, PlaybackDirection::backward);
+   const auto numSamplesReadRight = GetFloatsFromTime(
+      t, iChannel, buffer + numSideSamples, numSideSamples + 1, mayThrow,
+      PlaybackDirection::forward);
+   return { numSideSamples - numSamplesReadLeft,
+            numSideSamples + numSamplesReadRight };
+}
+
+namespace
+{
+template<typename BufferType>
+struct BufferCharType;
+
+template<>
+struct BufferCharType<float*>
+{
+   using type = samplePtr;
+};
+
+template<>
+struct BufferCharType<const float*>
+{
+   using type = constSamplePtr;
+};
+
+template <typename BufferType> struct SampleAccessArgs
+{
+   const typename BufferCharType<BufferType>::type offsetBuffer;
+   const sampleCount start;
+   const size_t len;
+};
+
+template <typename BufferType>
+SampleAccessArgs<BufferType> GetSampleAccessArgs(
+   const WaveClip& clip, double startOrEndTime /*absolute*/, BufferType buffer,
+   size_t totalToRead, size_t alreadyRead, bool forward)
+{
+   const auto remainingToRead = totalToRead - alreadyRead;
+   const auto sampsInClip = clip.GetVisibleSampleCount();
+   const auto sampsPerSec = clip.GetRate() / clip.GetStretchRatio();
+   if (forward)
+   {
+      const auto startTime =
+         std::max(startOrEndTime - clip.GetPlayStartTime(), 0.);
+      const sampleCount startSamp { std::ceil(
+         std::round(startTime * sampsPerSec * 1e6) / 1e6) };
```

Comment:
Explain the strange use of `1e6`

Reply:
Rounding to the nearest millionth of a sample?  Why?

Reply:
I can't remember. Just did a quick test with regular rounding, works too. I'll use a simple rounding, and next time should comment.